Regression with `@JsonCreator` that returns `null` for empty strings

### Search before asking

- [x] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

While updating to Jackson 3.0, our test suite revealed a regression introduced in #5090
where deserializing from a property with an empty string value using a `@JsonCreator` method that transforms this value to `null` causes the empty string to be assigned into the field as-is, raising `ClassCastException`.

### Version Information

3.0.1

### Reproduction

```java
    static class NonEmpty {
      public NonEmptyString nonEmpty;
    }

    static class NonEmptyString {
        public final String value;

        @JsonCreator
        public static NonEmptyString of(String value) {
            if (value == null || value.isEmpty()) {
                return null;
            }
            return new NonEmptyString(value);
        }

        private NonEmptyString(String value) {
            this.value = value;
        }
    }

    @Test
    void testDeserializeFieldToNullIfCreatorReturnsNull()
            throws Exception
    {
        NonEmpty result = MAPPER.readValue(
                "{ \"nonEmpty\": \"\" }",
                NonEmpty.class);

        assertNotNull(result);
        assertNull(result.nonEmpty);
    }
``` 

```
[ERROR] Errors: 
[ERROR]   JsonCreatorReturningNull4938Test.testDeserializeToNullWhenAllPropertyIsNull:119 ╗ Databind Cannot cast java.lang.String to tools.jackson.databind.deser.creators.JsonCreatorReturningNull4938Test$Localized1
 at [Source: (String)"{ "locale": "" }"; line: 1, column: 15] (through reference chain: tools.jackson.databind.deser.creators.JsonCreatorReturningNull4938Test$Localized["locale"])
```

### Expected behavior

A passing test, where `NonEmpty.nonEmpty` is `null`.

### Additional context

I will be opening a PR with fix shortly

--- a/src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java
+++ b/src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java
@@ -107,8 +107,7 @@ public final class AnnotatedMethod
     @Override
     public final Object call1(Object arg) throws Exception {
         try {
-            Object ret = _invokerUnary.get().invokeExact(arg);
-            return ret == null ? arg : ret;
+            return _invokerUnary.get().invokeExact(arg);
         } catch (final Throwable e) {
             throw sneakyThrow(e);
         }

--- a/src/test/java/tools/jackson/databind/deser/creators/JsonCreatorReturningNull4938Test.java
+++ b/src/test/java/tools/jackson/databind/deser/creators/JsonCreatorReturningNull4938Test.java
@@ -13,29 +13,11 @@ import tools.jackson.databind.testutil.DatabindTestUtil;
 import static org.junit.jupiter.api.Assertions.*;
 
 // [databind#4938] Allow JsonCreator factory method to return `null`
+// [databind#5401] Do not replace null from JsonCreator factory method
 public class JsonCreatorReturningNull4938Test
     extends DatabindTestUtil
 {
-    static class NonEmpty {
-        public NonEmptyString nonEmpty;
-    }
-
-    static class NonEmptyString {
-        public final String value;
-
-        @JsonCreator
-        public static NonEmptyString of(String value) {
-            if (value == null || value.isEmpty()) {
-                return null;
-            }
-            return new NonEmptyString(value);
-        }
-
-        private NonEmptyString(String value) {
-            this.value = value;
-        }
-    }
-
+    // [databind#4938]
     static class Localized3 {
         public final String en;
         public final String de;
@@ -109,24 +91,32 @@ public class JsonCreatorReturningNull4938Test
         }
     }
 
+    // [databind#5401]
+    static class NonEmpty5401 {
+        public NonEmptyString5401 nonEmpty;
+    }
+
+    static class NonEmptyString5401 {
+        public final String value;
+
+        @JsonCreator
+        public static NonEmptyString5401 of(String value) {
+            if (value == null || value.isEmpty()) {
+                return null;
+            }
+            return new NonEmptyString5401(value);
+        }
+
+        private NonEmptyString5401(String value) {
+            this.value = value;
+        }
+    }
 
     private final ObjectMapper MAPPER = newJsonMapper();
 
-    @Test
-    void testDeserializeFieldToNullIfCreatorReturnsNull()
-            throws Exception
-    {
-        NonEmpty result = MAPPER.readValue(
-                "{ \"nonEmpty\": \"\" }",
-                NonEmpty.class);
-
-        assertNotNull(result);
-        assertNull(result.nonEmpty);
-    }
-
+    // [databind#4938]
     @Test
     void testDeserializeToNullWhenAllPropertiesAreNull()
-            throws Exception
     {
         Localized3 result = MAPPER.readValue(
                 "{ \"en\": null, \"de\": null, \"fr\": null }",
@@ -137,7 +127,6 @@ public class JsonCreatorReturningNull4938Test
 
     @Test
     void testDeserializeToNonNullWhenAnyPropertyIsNonNull()
-            throws Exception
     {
         Localized3 result = MAPPER.readValue(
                 "{ \"en\": \"Hello\", \"de\": null, \"fr\": null }",
@@ -149,7 +138,6 @@ public class JsonCreatorReturningNull4938Test
 
     @Test
     void testDeserializeReadingAfterCreatorProps()
-            throws Exception
     {
         // Should all fail...
         ObjectReader enabled = MAPPER.readerFor(Localized4.class)
@@ -167,7 +155,6 @@ public class JsonCreatorReturningNull4938Test
     // Test to verify we are reading till the end of the OBJECT
     @Test
     void testDeserializeReadingUntilEndObject()
-            throws Exception
     {
         // Should all fail...
         ObjectReader enabled = MAPPER.readerFor(Localized4.class)
@@ -188,7 +175,6 @@ public class JsonCreatorReturningNull4938Test
 
     @Test
     void testJsonCreatorNullWithAnySetter()
-            throws Exception
     {
         String JSON = "{ \"en\": null, \"de\": null, \"fr\": null, " +
                 // These two properties are unknown
@@ -197,5 +183,17 @@ public class JsonCreatorReturningNull4938Test
         MAPPER.readerFor(Localized5.class)
                 .without(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
                 .readValue(JSON);
+    }
+
+    // [databind#5401]
+    @Test
+    void deserializeFieldToNullIfDelegatingCreatorReturnsNull()
+    {
+        NonEmpty5401 result = MAPPER.readValue(
+                "{ \"nonEmpty\": \"\" }",
+                NonEmpty5401.class);
+
+        assertNotNull(result);
+        assertNull(result.nonEmpty);
     }
 }

--- a/src/test/java/tools/jackson/databind/deser/creators/JsonCreatorReturningNull4938Test.java
+++ b/src/test/java/tools/jackson/databind/deser/creators/JsonCreatorReturningNull4938Test.java
@@ -16,6 +16,26 @@ import static org.junit.jupiter.api.Assertions.*;
 public class JsonCreatorReturningNull4938Test
     extends DatabindTestUtil
 {
+    static class NonEmpty {
+        public NonEmptyString nonEmpty;
+    }
+
+    static class NonEmptyString {
+        public final String value;
+
+        @JsonCreator
+        public static NonEmptyString of(String value) {
+            if (value == null || value.isEmpty()) {
+                return null;
+            }
+            return new NonEmptyString(value);
+        }
+
+        private NonEmptyString(String value) {
+            this.value = value;
+        }
+    }
+
     static class Localized3 {
         public final String en;
         public final String de;
@@ -91,6 +111,18 @@ public class JsonCreatorReturningNull4938Test
 
 
     private final ObjectMapper MAPPER = newJsonMapper();
+
+    @Test
+    void testDeserializeFieldToNullIfCreatorReturnsNull()
+            throws Exception
+    {
+        NonEmpty result = MAPPER.readValue(
+                "{ \"nonEmpty\": \"\" }",
+                NonEmpty.class);
+
+        assertNotNull(result);
+        assertNull(result.nonEmpty);
+    }
 
     @Test
     void testDeserializeToNullWhenAllPropertiesAreNull()

